components/suggestion/command_provider/command_parser.ts

```diff
@@ -0,0 +1,619 @@
+};
+
+export type TextElementProps = {
+	subtype: string;
+	min_length: number;
+	max_length: number;
+	// options - encoding, regexp, etc.
+}
+
+export type AutocompleteElementProps = {
+	flag_name: string;
+
+	hint: string;
+
+	role_id: string;
+	positional: boolean;
+};
+
+export type AutocompleteProps = ElementProps & AutocompleteElementProps;
+
+export type AutocompleteText = AutocompleteProps & TextElementProps;
+
+export type AutocompleteStaticSelect = AutocompleteProps & StaticSelectElementProps;
+
+export type AutocompleteDynamicSelect = AutocompleteProps & DynamicSelectElementProps;
+
+export type AutocompleteBool = AutocompleteProps;
+export type AutocompleteUser = AutocompleteProps;
+export type AutocompleteChannel = AutocompleteProps;
+
+export type AutocompleteElement = AutocompleteText | AutocompleteStaticSelect | AutocompleteDynamicSelect | AutocompleteBool | AutocompleteUser | AutocompleteChannel;
+
+export type SubCommand = {
+    app_id: string;
+    form_url?: string;
+    pretext: string;
+    fullPretext?: string;
+    description?: string;
+    args?: AutocompleteElement[];
+    sub_commands?: SubCommand[];
+}
+
+export type SuggestionChoice = {
+    suggestion: string;
+    complete?: string;
+    getComplete?: (pretext: string, value: string) => string;
+    description?: string;
+    hint?: string;
+    iconData?: string;
+}
+    // {
+    //     complete: '/' + sug.Complete,
+    //     suggestion: '/' + sug.Suggestion,
+    //     hint: sug.Hint,
+    //     description: sug.Description,
+    //     iconData: sug.IconData,
+    // }
+
+export const decorateSuggestionChoiceComplete = (pretext: string, choice: SuggestionChoice): SuggestionChoice => {
+    if (choice.complete?.endsWith(EXECUTE_CURRENT_COMMAND_ITEM_ID)) {
+        return choice;
+    }
+
+    const words = pretext.split(' ');
+    words.pop();
+    const before = words.join(' ') + ' ';
+
+    if (choice.getComplete) {
+        choice.complete = choice.getComplete(pretext, choice.suggestion);
+    }
+    choice.hint = choice.hint || '';
+
+    if (!words.length) {
+        // base command
+        return choice;
+    }
+
+    return {
+        ...choice,
+        suggestion: '/' + choice.suggestion,
+        complete: before + choice.suggestion,
+    };
+}
+
+const extractBaseCommand = (pretext: string) => {
+    return pretext.substring(1).split(' ')[0];
+}
+
+export const shouldHandleCommand = (pretext: string, locations: PluginLocation[]): boolean => {
+    for (const loc of locations) {
+        let trigger = loc.trigger;
+        if (trigger[0] !== '/') {
+            trigger = '/' + trigger;
+        }
+
+        if (pretext.startsWith(trigger + ' ')) {
+            return true;
+        }
+    }
+    return false;
+}
+
+let ARGS = {};
+let COMMANDS: SubCommand[] = [];
+export const setCommands = ((cmds: SubCommand[]) => COMMANDS = cmds);
+
+export const fetchCloudAppCommand = async (location: PluginLocation): Promise<SubCommand> => {
```

Comment:
In this file there are also many "Cloud App" references. They should just be App.